### PR TITLE
CDAP-3761: DBSink columns property is not used to create the DBRecord

### PR DESCRIPTION
**Pull request by jwang47**
- [x] Tested locally in standalone with MySQL server

### CODE REVIEW

File: cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java [hunk 1]
```diff
@@ -58,6 +58,15 @@ public Schema getSchema() {
   }
 
   /**
+   * Get the fields in the record.
+   *
+   * @return the fields in the record.
+   */
+  public Map<String, Object> getFields() {
```
Comment (gokulavasan) [inline cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java]: Do we need to add this method? Can't we get the schema and then get fields from it?

Comment (jwang47) [inline cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java]: I added this to make it easier to write the unit test.

Comment (bdmogal) [inline cdap-api-common/src/main/java/co/cask/cdap/api/data/format/StructuredRecord.java]: This is an internal detail. should probably not expose it outside, especially just for tests.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java [hunk 2]
```diff
@@ -141,11 +169,19 @@ private void setResultSetMetadata() throws Exception {
...
         // JDBC driver column indices start with 1
-        for (int i = 0; i < columnCount; i++) {
-          columnTypes[i] = resultSetMetadata.getColumnType(i + 1);
+        for (int i = 0; i < rs.getMetaData().getColumnCount(); i++) {
+          String name = resultSetMetadata.getColumnName(i + 1);
+          int type = resultSetMetadata.getColumnType(i + 1);
+          columnToType.put(name, type);
+        }
+
+        columns = ImmutableList.copyOf(Splitter.on(",").split(dbSinkConfig.columns));
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: can this (line 180-185) move outside the try-catch? Want to do as less as possible while the ``Connection is active.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java [hunk 3]
```diff
@@ -141,13 +170,20 @@ private void setResultSetMetadata() throws Exception {
...
-        for (int i = 0; i < columnCount; i++) {
-          columnTypes[i] = resultSetMetadata.getColumnType(i + 1);
+        for (int i = 0; i < rs.getMetaData().getColumnCount(); i++) {
+          String name = resultSetMetadata.getColumnName(i + 1);
+          int type = resultSetMetadata.getColumnType(i + 1);
+          columnToType.put(name, type);
         }
       }
+
+      columns = ImmutableList.copyOf(Splitter.on(",").split(dbSinkConfig.columns));
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: can it move outside this try-catch too?

I mean after the connection is closed?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java [hunk 4]
```diff
@@ -111,7 +118,20 @@ public void initialize(BatchRuntimeContext context) throws Exception {
 
   @Override
   public void transform(StructuredRecord input, Emitter<KeyValue<DBRecord, NullWritable>> emitter) throws Exception {
-    emitter.emit(new KeyValue<DBRecord, NullWritable>(new DBRecord(input, columnTypes), null));
+    // Create StructuredRecord that only has the columns in this.columns
+    List<Schema.Field> outputFields = Lists.newArrayList();
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: ``new ArrayList<>()``

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java [hunk 5]
```diff
@@ -111,7 +118,20 @@ public void initialize(BatchRuntimeContext context) throws Exception {
...
+    for (String column : getColumns()) {
+      Schema.Field field = input.getSchema().getField(column);
+      Preconditions.checkNotNull(field, "Missing schema field for column '%s'", column);
+      outputFields.add(field);
+    }
+    StructuredRecord.Builder output = StructuredRecord.builder(
+      Schema.recordOf(input.getSchema().getRecordName(), outputFields));
+    for (String column : getColumns()) {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: nit: can use ``columns`` directly here instead of a method call?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java [hunk 6]
```diff
@@ -111,7 +118,20 @@ public void initialize(BatchRuntimeContext context) throws Exception {
...
-    emitter.emit(new KeyValue<DBRecord, NullWritable>(new DBRecord(input, columnTypes), null));
+    // Create StructuredRecord that only has the columns in this.columns
+    List<Schema.Field> outputFields = Lists.newArrayList();
+    for (String column : getColumns()) {
+      Schema.Field field = input.getSchema().getField(column);
+      Preconditions.checkNotNull(field, "Missing schema field for column '%s'", column);
+      outputFields.add(field);
+    }
+    StructuredRecord.Builder output = StructuredRecord.builder(
+      Schema.recordOf(input.getSchema().getRecordName(), outputFields));
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: ``input.getSchema()`` will give the overall schema right? You only want a subset of it, consisting of ``columns`` only?

Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: nvm. I misread this.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java [hunk 7]
```diff
@@ -125,6 +145,14 @@ public void destroy() {
     DBUtils.cleanup(driverClass);
   }
 
+  public int[] getColumnTypes() {
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: do we need this to be exposed? Its only used in this class.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/batch/sink/DBSinkTest.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    MockEmitter<KeyValue<DBRecord, NullWritable>> emitter = new MockEmitter<>();
+    sink.transform(input, emitter);
+
+    Assert.assertEquals(1, emitter.getEmitted().size());
+    Assert.assertEquals(
+      getRecordFields(output),
+      getRecordFields(emitter.getEmitted().get(0).getKey().getRecord()));
+  }
+
+  @Test(expected = NullPointerException.class)
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/batch/sink/DBSinkTest.java]: nit: Where will this be thrown? If you are expecting a generic exception like ``NullPointerException`` (which could be thrown at multiple places, and not a specific exception, then it would be better to enclose the exact lines that throw ``NullPointerException`` in a try catch and ignore the exception.

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/batch/sink/DBSinkTest.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+        Schema.Field.of("body", Schema.of(Schema.Type.STRING)),
+        Schema.Field.of("ts", Schema.of(Schema.Type.LONG))))
+      .set("body", "sdfsdf")
+      .set("ts", 123)
+      .build();
+
+    MockEmitter<KeyValue<DBRecord, NullWritable>> emitter = new MockEmitter<>();
+    sink.transform(input, emitter);
+
+    Assert.assertEquals(1, emitter.getEmitted().size());
```
Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/test/java/co/cask/cdap/etl/batch/sink/DBSinkTest.java]: are these assertions needed if there is an exception? there shouldn't be anything in the emitter in this case, right?

File: cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java (hunk 5 above)
Comment (jwang47) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: This is useful for testing when we want to use a specific list of columns without having to go through unrelated code paths.

Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: ok, then lets make it package-private and ``@VisibleForTesting``? for both those methods?

Comment (bdmogal) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: also a comment perhaps, so people don't remove it later?

Comment (jwang47) [inline cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/etl/batch/sink/DBSink.java]: Seems like overkill, especially since breaking this would cause `DBSinkTest` to fail, which only takes 1s to run.

**Comment by bdmogal** [comment]
Code LGTM.

**Comment by jwang47** [comment]
There's an issue with how this handles Oracle SQL tables, because when we fetch the column info, the column names are always in uppercase, so our schema column names also need to be in uppercase. We want to provide some way to handle this, similar to how it's done in DBSource.

For now, we'll merge to develop first and fix the uppercase issue. Once it's fixed, we'll cherry-pick the commits to release/3.2.

**Comment by jwang47** [comment]
Reopened against develop: https://github.com/caskdata/cdap/pull/4149
